### PR TITLE
Improved Solver Interface Pass #1

### PR DESCRIPTION
**Pull request by hendrik-skubch**
Improves the solver interface by
- Introduce a variant type to be used in the result store
- Introduce a Blackboard for larger data that can be referred to by variants.
- Template over solver and result type to give Query some type safety back.
- various minor improvements
- necessary changes to supplementary are here: https://github.com/rapyuta-robotics/alica-supplementary/pull/7

### CODE REVIEW

File: alica_ros_proxy/src/communication/AlicaRosCommunication.cpp [hunk 1]
```diff
@@ -193,14 +193,14 @@ void AlicaRosCommunication::sendSyncTalk(SyncTalk& st) const {
...
     srs.senderID.id = sr.senderID->toByteVector();
 
-    for (auto sv : sr.vars) {
+    for (const SolverVar& sv : sr.vars) {
         alica_msgs::SolverVar svs;
-        svs.id = sv->id;
-        svs.value = sv->value;
+        svs.id = sv.id;
+        svs.value = std::vector<uint8_t>(sv.value, sv.value + sizeof(sv.value) / sizeof(sv.value[0]));
         srs.vars.push_back(svs);
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_ros_proxy/src/communication/AlicaRosCommunication.cpp]: nice catch!

File: alica_engine/src/engine/blackboard/ByteArray.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+ByteArray::ByteArray(ByteArray&& o)
+    : _begin(o._begin)
+    , _size(o._size) {
+        o._begin = nullptr;
+        //setting the size is not necessary
+}
+
+ByteArray& ByteArray::operator=(ByteArray&& o) {
+    std::swap(_begin,o._begin);
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/src/engine/blackboard/ByteArray.cpp]: common practice, this is a move operator (see the rvalue argument)

File: alica_engine/src/engine/blackboard/ByteArray.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+namespace alica {
+ByteArray::ByteArray(const int8_t* src, int32_t count)
+    : _size(count) {
+    assert(src != nullptr);
+    if (count > 0) {
+        _begin = static_cast<int8_t*>(malloc(count));
+        assert(_begin!=nullptr);
+        memcpy(_begin, src, count);
+    } else {
+        _begin = nullptr;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/src/engine/blackboard/ByteArray.cpp]: done

File: alica_engine/include/engine/blackboard/BlackBoard.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    IdType registerValue(const int8_t* buffer, int len);
+    IdType registerValue(const char* buffer, int len);
+    
+    template <class InputIt>
+    IdType registerValue(InputIt begin, InputIt end);
+
+    bool hasValue(IdType id) const {return _body.find(id) != _body.end();}
+
+    const ObjectType& getValue(IdType id) const {
+        return _body.find(id)->second;
```
[the remark this answers is not in the document]
Comment (hendrik-skubch) [reply alica_engine/include/engine/blackboard/BlackBoard.h]: done

File: alica_engine/src/engine/blackboard/ByteArray.cpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    }
+}
+
+ByteArray::ByteArray(int32_t size)
+    : _size(size)
+{
+    assert(size > 0);
+    if(size> 0) {
+        _begin = static_cast<int8_t*>(malloc(size));
+    } else {
```
Comment (gauthamm) [inline alica_engine/src/engine/blackboard/ByteArray.cpp]: I know this PR is merged already but I just observed that this else is redundant.

**Pull request merged**